audb/core/stream.py

```diff
@@ -0,0 +1,701 @@
+
+        >>> next(db)
+                                         duration  speaker transcription
+        file
+        wav/03a01Fa.wav    0 days 00:00:01.898250        3           a01
+        wav/03a01Nc.wav    0 days 00:00:01.611250        3           a01
+        wav/03a01Wa.wav 0 days 00:00:01.877812500        3           a01
+        wav/03a02Fc.wav    0 days 00:00:02.006250        3           a02
+
+        During the iteration,
+        the :class:`audb.DatabaseIterator` object
+        provides access to the current batch of data.
+
+        >>> db["files"].get(map={"speaker": "age"})
+                                         duration transcription  age
+        file
+        wav/03a01Fa.wav    0 days 00:00:01.898250           a01   31
+        wav/03a01Nc.wav    0 days 00:00:01.611250           a01   31
+        wav/03a01Wa.wav 0 days 00:00:01.877812500           a01   31
+        wav/03a02Fc.wav    0 days 00:00:02.006250           a02   31
+
+    """  # noqa: E501
+
+    def __init__(
+        self,
+        db: audformat.Database,
+        table: str,
+        *,
+        version: str,
+        map: typing.Dict[str, typing.Union[str, typing.Sequence[str]]],
+        batch_size: int,
+        shuffle: bool,
+        buffer_size: int,
+        only_metadata: bool,
+        bit_depth: int,
+        channels: typing.Union[int, typing.Sequence[int]],
+        format: str,
+        mixdown: bool,
+        sampling_rate: int,
+        full_path: bool,
+        cache_root: str,
+        num_workers: typing.Optional[int],
+        timeout: float,
+        verbose: bool,
+    ):
+        self._cleanup_database(db, table)
+
+        # Transfer attributes of database object
+        for attr in db.__dict__.keys():
+            setattr(self, attr, getattr(db, attr))
+
+        self._table = table
+        self._version = version
+        self._map = map
+        self._batch_size = batch_size
+        self._shuffle = shuffle
+        self._buffer_size = buffer_size
+        self._only_metadata = only_metadata
+        self._bit_depth = bit_depth
+        self._channels = channels
+        self._format = format
+        self._mixdown = mixdown
+        self._sampling_rate = sampling_rate
+        self._full_path = full_path
+        self._cache_root = cache_root
+        self._num_workers = num_workers
+        self._timeout = timeout
+        self._verbose = verbose
+
+        if shuffle:
+            self._samples = buffer_size
+        else:
+            self._samples = batch_size
+        self._buffer = pd.DataFrame()
+        self._stream = self._initialize_stream()
+
+    def __iter__(self) -> DatabaseIterator:
+        r"""Iterator generator."""
+        return self
+
+    def __next__(self) -> pd.DataFrame:
+        r"""Iterate database."""
+        # Load part of table
+        df = self._get_batch()
+        self[self._table]._df = df
+
+        # Load corresponding media files
+        self._load_media(df)
```

Comment:
This implements `next` of the iterator protocol. 

I wonder about the behavior / effect of  calling `_load_media` here.

Would this not result in throttling streaming on the first call?
And when setting to shuffle even on further iterations as more media would have to be pulled into the cache first before the batch becomes available. 

If this is correct would it not be possible to (optionally) prefetch media?

Reply:
> Would this not result in throttling streaming on the first call?

It shouldn't, as it only has to load the media file requested by the batch (assuming that we have the data published with one archive per file), e.g. when we have a database with 100,000 files, and we use `batch_size=16` it will only download 16 media files with the first `next(db)` call, then another 16 media files with the next call, ...

Shuffling of the data has no consequences for loading the media, it only requires to load `buffer_size` number of lines from the table instead of `batch_size`, but the media files to be downloaded per iteration is still given by `batch_size`.

If you want to prefetch all media files, you can use `audb.load()` or `audb.load_media()` (the later if you don't want to load the tables).

Reply:
Prefetching can always be done calling `audb.load()` or `audb.load_to()`.  But wouldn't that run counter the description:

>> allows downloading media files of databases, that do not fit onto hard disc


Of course I do not know enough about the interaction of  `batch_size`, segmented indices file sizes.
I would have thought that a typical usecase that someone might use this for streaming data in model training. Hence the request.  I was thinking in terms of training batch sizes to become 64 or 128.

Reply:
> Prefetching can always be done calling audb.load() or audb.load_to(). But wouldn't that run counter the description:
>
>> allows downloading media files of databases, that do not fit onto hard disc

Yes, of cause. If a database does not fit onto your hard disc, you should not prefetch it.

> I was thinking in terms of training batch sizes to become 64 or 128.

Downloading 128 media files on the fly takes not that long. And if speed is an issue, you can still pre-fetch the data, as at the moment we don't have any database that does fit into memory. But streaming also allows for a very easy preview of the database. Before a user had to use `audb.load(..., only_metadata)` select interesting media files and then run `audb.load(..., media=media)`. This can now be replaced by a single `audb.stream()` call.

Reply:
Alright then I see. Then even with a larget batch this will not be problematic. 
And a very prominent usecase it the more handy preview of databases. 

As this was then owed only to my limited knowledge abot usecases etc. I will resolve this conversation.